librad/src/identities/git.rs

```diff
@@ -0,0 +1,763 @@
+    /// the verification succeeded -- which may or may not be `head`.
+    pub fn verify(&self, head: git2::Oid) -> Result<VerifiedUser, error::VerifyUser> {
+        Ok(self.verify_generic(head)?)
+    }
+
+    /// Create a new [`User`] from a payload and delegations.
+    ///
+    /// The returned [`User`] (and the underlying commit) will not have any
+    /// parents, and will by signed by `signer`.
+    pub fn create<S>(
+        &self,
+        payload: UserPayload,
+        delegations: delegation::Direct,
+        signer: &S,
+    ) -> Result<User, error::Store<S::Error>>
+    where
+        S: Signer,
+        S::Error: std::error::Error + Send + Sync + 'static,
+    {
+        let doc = Doc {
+            version: 0,
+            replaces: None,
+            payload,
+            delegations: payload::UserDelegations::from(delegations),
+        };
+
+        let root: Revision = self.repo.blob(&Cjson(&doc).canonical_form()?)?.into();
+        let revision = {
+            let mut builder = self.repo.treebuilder(None)?;
+            builder.insert(root.to_string(), *root, 0o100_644)?;
+            builder.write().map(Revision::from)
+        }?;
+        let signatures = sign(signer, revision).map_err(error::Store::Signer)?.into();
+        let content_id = self.commit(
+            &format!("Initialised user identity {}", root),
+            &signatures,
+            revision,
+            &[],
+        )?;
+
+        Ok(Identity {
+            content_id,
+            root,
+            revision,
+            doc: doc.second(delegation::Direct::from),
+            signatures,
+        })
+    }
+
+    /// Update an existing [`SignedUser`] with a new payload and delegations.
+    ///
+    /// If both `payload` and `delegations` evaluate to `None`, or their values
+    /// result in the same revision as `base`, no new commit is made, and
+    /// the result is the unwrapped [`User`] of the `base` argument.
+    ///
+    /// Otherwise, the result is a new [`User`] whose parent is `base`.
+    pub fn update<S>(
+        &self,
+        base: SignedUser,
+        payload: impl Into<Option<UserPayload>>,
+        delegations: impl Into<Option<delegation::Direct>>,
+        signer: &S,
+    ) -> Result<User, error::Store<S::Error>>
+    where
+        S: Signer,
+        S::Error: std::error::Error + Send + Sync + 'static,
+    {
+        let payload = payload.into();
+        let delegations = delegations.into();
+
+        // Fast path
+        if payload.is_none() && delegations.is_none() {
+            return Ok(base.into_inner());
+        }
+
+        let doc = Doc {
+            version: 0,
+            replaces: Some(base.revision),
+            payload: payload.unwrap_or_else(|| base.doc.payload.clone()),
+            delegations: delegations
+                .or_else(|| Some(base.doc.delegations.clone()))
+                .map(payload::UserDelegations::from)
+                .unwrap(),
```

Comment:
I'm a little confused by this chain. Why does this end up being safe to `unwrap`?

Reply:
Because `or_else` makes it always `Some`. Can do a pattern match if you prefer.